lib/fdtable.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief File descriptor table
+ *
+ * This file provides generic file descriptor table implementation, suitable
+ * for any I/O object implementing POSIX I/O semantics (i.e. read/write +
+ * aux operations).
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <misc/fdtable.h>
+
+struct fd_entry {
+	void *obj;
+	const struct fd_op_vtable *vtable;
+};
+
+#ifdef CONFIG_POSIX_API
+static struct fd_op_vtable stdinout_fd_op_vtable;
+#endif
+
+static struct fd_entry fdtable[CONFIG_POSIX_MAX_FDS] = {
+#ifdef CONFIG_POSIX_API
+	/*
+	 * Predefine entries for stdin/stdout/stderr. Object pointer
+	 * is unused and just should be !0 (random different values
+	 * are used to posisbly help with debugging).
+	 */
+	{(void *)0x10, &stdinout_fd_op_vtable},
+	{(void *)0x11, &stdinout_fd_op_vtable},
+	{(void *)0x12, &stdinout_fd_op_vtable},
+#endif
+};
+
+static K_MUTEX_DEFINE(fdtable_lock);
+
+static int _find_fd_entry(void)
+{
+	int fd;
+
+	for (fd = 0; fd < ARRAY_SIZE(fdtable); fd++) {
+		if (fdtable[fd].obj == NULL) {
+			return fd;
+		}
+	}
+
+	errno = ENFILE;
+	return -1;
+}
+
+static int _check_fd(int fd)
+{
+	if (fd < 0 || fd >= ARRAY_SIZE(fdtable) || fdtable[fd].obj == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	return 0;
+}
+
+void *z_get_fd_obj(int fd, const struct fd_op_vtable *vtable, int err)
+{
+	struct fd_entry *fd_entry;
+
+	if (_check_fd(fd) < 0) {
+		return NULL;
+	}
+
+	fd_entry = &fdtable[fd];
+
+	if (vtable != NULL && fd_entry->vtable != vtable) {
+		errno = err;
+		return NULL;
+	}
+
+	return fd_entry->obj;
+}
+
+int z_reserve_fd(void)
+{
+	int fd;
+
+	(void)k_mutex_lock(&fdtable_lock, K_FOREVER);
+
+	fd = _find_fd_entry();
+
+	if (fd >= 0) {
+		/* Mark entry as used, z_finalize_fd() will fill it in. */
+		fdtable[fd].obj = (void *)1;
```

Comment:
I would rather see a define used here instead of magic constant.

Reply:
Ok.